backend/crates/eiffelvis_core/src/domain/user_queries.rs

```diff
@@ -0,0 +1,105 @@
+use crate::{
+    domain::types::BaseEvent,
+    graph::*,
+    tracked_query::{TrackedNodes, TrackedSubGraphs},
+};
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+/// Describes a query to collect nodes from an eiffel graph,
+/// to match an event **all** filters need to match.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Query {
+    /// Will return a given node if any of the filters match
+    filters: Vec<Filter>,
+    /// Method of collection, may add additional nodes
+    collection: Collection,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(tag = "type")]
+pub enum Filter {
+    /// Do not filter at all, just give everything
+    None,
+    /// Open ended range of time
+    Time {
+        begin: Option<u64>,
+        end: Option<u64>,
+    },
+    /// Event Type
+    Type { name: String },
+    /// Specific ids
+    Ids { ids: Vec<Uuid> },
+}
+
+/// Used collection method,
+/// selected variant is run **after** filtering has been done,
+/// this means you can get nodes back that may not filfull the filter requirements.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(tag = "type")]
+pub enum Collection {
+    /// Does not do anything
+    Forward,
+    /// Uses the result of the filtered nodes as roots and collects their graph
+    AsRoots,
+}
+
+/// A tracked query, only returns new matches.
+pub struct TrackedQuery<I> {
+    filters: Vec<Filter>,
+    collector: Collector<I>,
+    inner: TrackedNodes<I>,
+}
+
+enum Collector<I> {
+    Forward,
+    SubGraph(TrackedSubGraphs<I>),
+}
+
+impl<I> TrackedQuery<I> {
+    pub fn new(query: Query) -> Self {
+        Self {
+            filters: query.filters,
+            collector: match query.collection {
+                Collection::Forward => Collector::Forward,
+                Collection::AsRoots => Collector::SubGraph(TrackedSubGraphs::new(vec![])),
+            },
+            inner: TrackedNodes::new(),
+        }
+    }
+
+    /// Collects the new nodes since the last time this method was called that match the query.
```

Comment:
'Collects new nodes that match the query since the last time this method was called'